drivers/can/mcp2515.c

```diff
@@ -0,0 +1,651 @@
+		target[MCP2515_FRAME_OFFSET_EID0] = source->ext_id;
+	}
+
+	rtr = (source->rtr == CAN_REMOTEREQUEST) ? BIT(6) : 0;
+	dlc = (source->dlc) & 0x0F;
+
+	target[MCP2515_FRAME_OFFSET_DLC] = rtr | dlc;
+
+	for (; data_idx < 8; data_idx++) {
+		target[MCP2515_FRAME_OFFSET_D0 + data_idx] =
+			source->data[data_idx];
+	}
+}
+
+static void mcp2515_convert_mcp2515frame_to_zcanframe(const u8_t *source,
+						      struct zcan_frame *target)
+{
+	u8_t data_idx = 0U;
+
+	if (source[MCP2515_FRAME_OFFSET_SIDL] & BIT(3)) {
+		target->id_type = CAN_EXTENDED_IDENTIFIER;
+		target->ext_id =
+			(source[MCP2515_FRAME_OFFSET_SIDH] << 21) |
+			((source[MCP2515_FRAME_OFFSET_SIDL] >> 5) << 18) |
+			((source[MCP2515_FRAME_OFFSET_SIDL] & 0x03) << 16) |
+			(source[MCP2515_FRAME_OFFSET_EID8] << 8) |
+			source[MCP2515_FRAME_OFFSET_EID0];
+	} else {
+		target->id_type = CAN_STANDARD_IDENTIFIER;
+		target->std_id = (source[MCP2515_FRAME_OFFSET_SIDH] << 3) |
+				 (source[MCP2515_FRAME_OFFSET_SIDL] >> 5);
+	}
+
+	target->dlc = source[MCP2515_FRAME_OFFSET_DLC] & 0x0F;
+	target->rtr = source[MCP2515_FRAME_OFFSET_DLC] & BIT(6) ?
+		      CAN_REMOTEREQUEST : CAN_DATAFRAME;
+
+	for (; data_idx < 8; data_idx++) {
+		target->data[data_idx] = source[MCP2515_FRAME_OFFSET_D0 +
+						data_idx];
+	}
+}
+
+const int mcp2515_set_mode(struct device *dev, u8_t mcp2515_mode)
+{
+	u8_t canstat;
+
+	mcp2515_cmd_bit_modify(dev, MCP2515_ADDR_CANCTRL,
+			       MCP2515_CANCTRL_MODE_MASK,
+			       mcp2515_mode << MCP2515_CANCTRL_MODE_POS);
+	mcp2515_cmd_read_reg(dev, MCP2515_ADDR_CANSTAT, &canstat, 1);
+
+	if (((canstat & MCP2515_CANSTAT_MODE_MASK) >> MCP2515_CANSTAT_MODE_POS)
+	    != mcp2515_mode) {
+		LOG_ERR("Failed to set MCP2515 operation mode");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int mcp2515_attach(struct device *dev, const struct zcan_filter *filter,
+			  void *response_ptr, u8_t is_type_msgq)
+{
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+	int filter_idx = 0;
+
+	__ASSERT(response_ptr != NULL, "response_ptr can not be null");
+
+	k_mutex_lock(&dev_data->filter_mutex, K_FOREVER);
+
+	/* find free filter */
+	while ((BIT(filter_idx) & dev_data->filter_usage)
+	       && (filter_idx < CONFIG_CAN_MCP2515_MAX_FILTER)) {
+		filter_idx++;
+	}
+
+	/* setup filter */
+	if (filter_idx < CONFIG_CAN_MCP2515_MAX_FILTER) {
+		dev_data->filter_usage |= BIT(filter_idx);
+		if (is_type_msgq) {
+			dev_data->filter_response_type |= BIT(filter_idx);
+		} else {
+			dev_data->filter_response_type &= ~BIT(filter_idx);
+		}
+		dev_data->filter[filter_idx] = *filter;
+		dev_data->filter_response[filter_idx] = response_ptr;
+	} else {
+		filter_idx = CAN_NO_FREE_FILTER;
+	}
+
+	k_mutex_unlock(&dev_data->filter_mutex);
+
+	return filter_idx;
+}
+
+static int mcp2515_configure(struct device *dev, enum can_mode mode,
+			     u32_t bitrate)
+{
+	const struct mcp2515_config *dev_cfg = DEV_CFG(dev);
+
+	/* CNF3, CNF2, CNF1, CANINTE */
+	u8_t config_buf[4];
+
+	if (bitrate == 0) {
```

Comment:
Maybe we should make it more explicit somewhere, that bus-speed confgirued in device-tree will be ignored, if can_configure is called with `bitrate != 0` ?

Reply:
Indeed. This should be added in the API documentation.

Reply:
haha actually I only found this because the api test tripped there and added it in after double checking with the stm32 can drivers :-)